Application/EdFi.Ods.Features/Profiles/ProfileMetadataDatabaseProvider.cs

```diff
@@ -0,0 +1,192 @@
+
+        private readonly IUsersContextFactory _usersContextFactory;
+        private readonly ICacheProvider _cacheProvider;
+        private readonly IProfileMetadataValidator _profileMetadataValidator;
+
+        private readonly ILog _logger = LogManager.GetLogger(typeof(ProfileMetadataDatabaseProvider));
+
+        public ProfileMetadataDatabaseProvider(
+            IUsersContextFactory usersContextFactory,
+            ICacheProvider cacheProvider,
+            IProfileMetadataValidator profileMetadataValidator)
+        {
+            _usersContextFactory = Preconditions.ThrowIfNull(usersContextFactory, nameof(usersContextFactory));
+            _cacheProvider = Preconditions.ThrowIfNull(cacheProvider, nameof(cacheProvider));
+            _profileMetadataValidator = Preconditions.ThrowIfNull(profileMetadataValidator, nameof(profileMetadataValidator));
+        }
+
+        bool IProfileMetadataProvider.ContainsProfileDefinition(string profileName)
+        {
+            EnsureProfileDefinitionsCacheIsRefreshedFromDatabase();
+
+            return _cacheProvider.TryGetCachedObject($"{DefinitionsCacheKeyPrefix}.{profileName.ToLowerInvariant()}", out object _);
+        }
+
+        XElement IProfileMetadataProvider.GetProfileDefinition(string profileName)
+        {
+            EnsureProfileDefinitionsCacheIsRefreshedFromDatabase();
+
+            if (!_cacheProvider.TryGetCachedObject($"{DefinitionsCacheKeyPrefix}.{profileName.ToLowerInvariant()}", out XElement profileDefinition))
+            {
+                return null;
+            }
+
+            return profileDefinition;
+        }
+
+        MetadataValidationResult[] IProfileMetadataProvider.GetValidationResults()
+        {
+            if (!_cacheProvider.TryGetCachedObject($"{ValidationsCacheKeyPrefix}", out IDictionary<(string name, string source), ValidationResult> validationResults))
+            {
+                return new MetadataValidationResult[] { };
+            }
+
+            return validationResults
+                .Select(kvp => new MetadataValidationResult(kvp.Key.name, kvp.Key.source, kvp.Value))
+                .ToArray();
+        }
+
+        List<ProfileAndResourceNames> IProfileResourceNamesProvider.GetProfileResourceNames()
+        {
+            EnsureProfileDefinitionsCacheIsRefreshedFromDatabase();
+
+            if (!_cacheProvider.TryGetCachedObject($"{ResourcesCacheKeyPrefix}", out List<ProfileAndResourceNames> profileResouces))
+            {
+                return new();
+            }
+
+            return profileResouces;
+        }
+
+        private void EnsureProfileDefinitionsCacheIsRefreshedFromDatabase()
+        {
+            if (!HasCacheExpired)
+            {
+                return;
+            }
+
+            // Add Active key to cache
+            _cacheProvider.SetCachedObject(CacheActiveKey, true);
+
+            List<Profile> profiles;
+            using (var usersContext = _usersContextFactory.CreateContext())
```

Comment:
Can you create new method GetProfiles in https://github.com/Ed-Fi-Alliance-OSS/Ed-Fi-ODS/blob/main/Application/EdFi.Admin.DataAccess/Repositories/ClientAppRepo.cs  and then inject IClientAppRepo interface to call GetProfiles method here instead of using DBContext in this file  .  Example here https://github.com/Ed-Fi-Alliance-OSS/Ed-Fi-ODS-Implementation/blob/main/Application/EdFi.Ods.Api.IntegrationTestHarness/UpdateAdminDatabaseTask.cs#L74